Added -help option to sourcekitd-test

felix91gr opened the pull request:
<!-- What's in this pull request? -->
### What the PR includes
Addresses one of the current starter bugs, adding the feature mentioned in the title.

I've added a `FIXME:` in the **printHelp** function to suggest a possible expansion for its functionality (allowing to hide/include certain options from the Help Text). I don't think that's necessary for this command, but could be useful if we want it to have as many options as `Driver.cpp` has. 

### Where has this been tested

I've compiled this in Linux, using a slightly modified version of the default presets. I used #8485 and #8189, and added a `cmake` option to compile `sourcekitd-test`. It passed the **buildbot_linux, smoketest** tests on my pc :+1: 

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-1421](https://bugs.swift.org/browse/SR-1421).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

felix91gr committed:
Added -help option so sourcekitd-test

Added "-help" option to Options.td
Added also:
* Added "OPT_HELP" case in the main TestOptions.cpp switch. It uses the llvm options help functionality to provide up-to-date help
* Additional "Use -help for assistance" at the end of the error message that appears when calling an unknown option
Added their following tests, respectively::
* usage.swift, and
* wrong_arguments.swift

Extra: FIXME: in TestOptions::printHelp, suggesting a possible expansion for the printHelp option (details in file)

felix91gr commented:
@swift-ci please smoke test

slavapestov commented:
@swift-ci Please test

nkcsgexi commented:
Can we add a test to ensure this dump the -h option itself?

felix91gr commented:
🤔 sure... I can add that

felix91gr commented:
Do we have a documentation on the testing tools? I'm talking about the kind of tests found in swift/test

felix91gr commented:
@nkcsgexi I made this test and it passed. Should I test something else as well?
```swift

// Basic check that usage works
//
// RUN: %sourcekitd-test -help | %FileCheck %s
//
// CHECK: USAGE: sourcekitd-test [options] <inputs>
```

nkcsgexi commented:
Thank you! @swift-ci please smoke test

felix91gr commented:
Thank you too🙂

Btw... 🤔 I made a new test, for checking that entering an unknown argument
prints an error message and suggests using -help. But running the 9xxx
tests takes a long time 😆.

Is there a way to run just *one* test? Or to run just the SourceKit ones?

On Tue, Apr 18, 2017 at 6:43 PM Xi Ge <notifications@github.com> wrote:

> *@nkcsgexi* approved this pull request.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/8746#pullrequestreview-33359112>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALNBJ5KmSGq0cNxpoH7psIDS40BgkWTHks5rxS6AgaJpZM4M9JRu>
> .
>

nkcsgexi commented:
@felix91gr could you push your test and ask swift CI to test it for you?

felix91gr commented:
Sure! I'll push it 🙂

I can't trigger swift CI though... I'm a new-ish member of the community,
so I haven't been given access to it.

(I already tried in my first comment, that was when I learned)

On Tue, Apr 18, 2017 at 7:30 PM Xi Ge <notifications@github.com> wrote:

> @felix91gr <https://github.com/felix91gr> could you push your test and
> ask swift CI to test it for you?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/8746#issuecomment-295002853>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALNBJ_ww0b1PB65nVTsiOjPHDbKHVv5Mks5rxTmagaJpZM4M9JRu>
> .
>

nkcsgexi commented:
@swift-ci please smoke test

felix91gr commented:
> please smoke test

Thanks! ^^

nkcsgexi commented:
@felix91gr whenever you need to trigger CI, just comment on the pull request to ask some one to trigger it for you. I'm sure many are glad to help 🙂

felix91gr commented:
It failed, this is the output:
```console
Command Output (stderr):
--
error: unknown argument: -this_option_does_not_exist
Use -h or -help for assistance
FileCheck error: '-' is empty.
FileCheck command line:  /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/llvm-macosx-x86_64/./bin/FileCheck /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/SourceKit/Misc/wrong_arguments.swift

--
```
From that, only the first two lines are the output of `sourcekitd-test`. That's the correct output (what we are testing, in fact). I have problems interpreting the `FileCheck` error though... what does...

 `FileCheck error: '-' is empty.` mean?

This is the original test, btw:

```swift
// RUN: %sourcekitd-test -this_option_does_not_exist | %FileCheck %s
//
// CHECK: error: unknown argument:
// CHECK: Use -h or -help for assistance
```

Should I use the complete line? (like `// CHECK: error: unknown argument: -this_option_does_not_exist`)

I thought `CHECK` didn't enforce the line to be exactly equal... I wanted to give freedom to have any erroneous option in the test, just in case. 

### (Meta)

:thinking: it failed only on OS X... Now that I look at it, the entire SourceKit test suite is disabled in Linux.

Ah, _right_. We haven't yet merged Norio Nomura's PR #8485. Well, I'll test it on my machine as well then (I'm using Linux) :+1:

felix91gr commented:
Yup, the `CHECK` string must be matched exactly. That must've been the error.

@nkcsgexi I've submitted a correction. Can you call Swift-CI please? :)

nkcsgexi commented:
@swift-ci please smoke test

felix91gr commented:
:thinking: both failed... that's weird.

Well, I'll restart my setup and keep at it.

felix91gr commented:
@benlangmuir or @nkcsgexi, I think I fixed the tests this time... sorry to bother you, could you run a smoke test please? :)

nkcsgexi commented:
@swift-ci please smoke test

felix91gr commented:
@benlangmuir or @nkcsgexi... Sorry to bother you again. This time I've tested the changes using a local version of `FileCheck`. It _should_ work. If it doesn't... :thinking: we'll see.

nkcsgexi commented:
@swift-ci please smoke test

felix91gr commented:
Wohoo! The `wrong_arguments.swift` test passed! I'm making it work :D 

Next up, I'll fix whatever is wrong with the `usage.swift` test :fist_raised:

felix91gr commented:
I think I know what's the problem. I'll upload the change.

felix91gr commented:
The problem was that I thought the `-help` command returned 0, because it doesn't trigger an error. **I was wrong**. 

In the design of `sourcekitd-test`, **any option** that makes the program not test a SourceKit request, **returns 1**. If the **"happy path"** would be to parse a request and perform a SourceKit test on it, then **anything but the happy path returns 1**. Therefore, both of my tests (`wrong_arguments` and `usage` should expect a 1 for result, and therefore should start with `{// RUN: not sourcekitd-test ...}`. Which now, they do :) _(I tried it first with `wrong_arguments`, and it passes)_

So, for (hopefully) last time... @nkcsgexi could you call CI? :)

nkcsgexi commented:
@swift-ci please smoke test

felix91gr commented:
Aw yeah! It passed! :D

felix91gr commented:
It passed! 
Thanks for your patience, @nkcsgexi :smile: 

It was a dumb mistake, as it always is :D

felix91gr commented:
Should we merge this? Should I add something else? 🙂

nkcsgexi commented:
Could you squash all commits into one?

felix91gr commented:
Oh, good idea. Will do.
On Fri, Apr 21, 2017 at 2:54 PM Xi Ge <notifications@github.com> wrote:

> Could you squash all commits into one?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apple/swift/pull/8746#issuecomment-296259391>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALNBJ70iBsgGIJ7KgxKLOnUr4BDMCeFWks5ryO08gaJpZM4M9JRu>
> .
>

felix91gr commented:
Done! :)

nkcsgexi commented:
👍 @swift-ci please smoke test and merge

felix91gr commented:
🤔 seems like nothing happened...

nkcsgexi commented:
Let's try again. @swift-ci please smoke test

felix91gr commented:
Ahhh... you know, probably when asked for a test + merge process, it puts the task way back into the queue because the tasks before it where called using the previous state of the repo :D 

If that's the case, then it should take a while sometimes, depending on how busy CI is :thinking:

akyrtzi commented on utils/build-presets.ini:
```diff
@@ -708,7 +708,7 @@ install-llbuild
 install-xctest
 install-prefix=/usr
-swift-install-components=autolink-driver;compiler;clang-builtin-headers;stdlib;swift-remote-mirror;sdk-overlay;license
+swift-install-components=autolink-driver;compiler;clang-builtin-headers;stdlib;swift-remote-mirror;sdk-overlay;license;sourcekit-inproc
```
This seems irrelevant, is this change intentional ?

felix91gr replied:
Oh. I didn't notice. It's a leftover of one of my cherrypicks from Nomura's PR (#8485). Helped me test things in linux.
It can, and probably should, be removed... even though that PR has already been aprooved. You can do it if you want, or just wait for me (atm I'm not at the computer) :)

akyrtzi replied:
Remove the change when you can, and afterwards we can merge your improvements.

felix91gr replied:
Oki doke 🙂👍🏻

felix91gr commented:
@akyrtzi done :wink:

felix91gr commented:
#### Updated
* I deleted everything that was not needed:
  * A `//WIP` comment that I made in `TestOptions.cpp`, and 
  * The `;sourcekit-inproc` install option in the build presets that I used for testing. 
* I also added:
  * A brief documentation on the new tests, and 
  * A _newline_ that was missing at the end of `TestOptions.cpp`

I think we're ready for a test and merge. @nkcsgexi or @akyrtzi could you please call CI again? :)

akyrtzi commented:
@swift-ci smoke test and merge

akyrtzi commented:
@swift-ci smoke test

felix91gr commented:
What a weird thing... why did it fail the [smoke test & merge](https://ci.swift.org/view/Pull%20Request/job/swift-PR-smoke-merge/2615/console), but passed the smoke test alone?

The only thing I can think of is that I haven't set gpg keys for my Github account. That could be it :thinking:

felix91gr commented:
No, wait... it says "wrong SHA". :thinking:
What made the SHA change?

felix91gr commented:
:smile: 

Thanks for the help and the patience, guys! @nkcsgexi and @akyrtzi specially :)

nkcsgexi commented:
@felix91gr congrats for your first Swift contribution! 👍
